api/include/opentelemetry/trace/propagation/b3_propagator.h

```diff
@@ -0,0 +1,278 @@
+      if (!(string_view[i] >= '0' && string_view[i] <= '9') &&
+          !(string_view[i] >= 'a' && string_view[i] <= 'f'))
+        return false;
+    }
+    return true;
+  }
+
+  static SpanContext ExtractImpl(Getter getter, const T &carrier)
+  {
+    // all these are hex values
+    nostd::string_view trace_id;
+    nostd::string_view span_id;
+    nostd::string_view trace_flags;
+
+    // first let's try a single-header variant
+    auto singleB3Header = getter(carrier, B3CombinedHeader);
+    if (singleB3Header != "")
+    {
+      trace_id = singleB3Header.substr(0, kTraceIdHexStrLength);
+      if (singleB3Header.size() > 33)
+      {
+        span_id = singleB3Header.substr(33, kSpanIdHexStrLength);
+      }
+      if (singleB3Header.size() > 50)
+      {
+        trace_flags = singleB3Header.substr(50, kTraceFlagHexStrLength);
+      }
+    }
+    else
+    {
+      trace_id    = getter(carrier, B3TraceIdHeader);
+      span_id     = getter(carrier, B3SpanIdHeader);
+      trace_flags = getter(carrier, B3SampledHeader);
+    }
+
+    // validate ids
+    if (!IsValidHex(trace_id) || !IsValidHex(span_id) || !IsValidHex(trace_flags) ||
+        trace_id == "00000000000000000000000000000000" || span_id == "0000000000000000")
+    {
+      return SpanContext(false, false);
+    }
+
+    // now convert hex to objects
+    TraceId trace_id_obj       = GenerateTraceIdFromString(trace_id);
+    SpanId span_id_obj         = GenerateSpanIdFromString(span_id);
+    TraceFlags trace_flags_obj = GenerateTraceFlagsFromString(trace_flags);
+    return SpanContext(trace_id_obj, span_id_obj, trace_flags_obj, true);
+  }
+};
+
+// The B3Propagator class provides an interface that enables extracting and injecting
+// context into headers of HTTP requests. HTTP frameworks and clients
+// can integrate with B3Propagator by providing the object containing the
+// headers, and a getter and setter function for the extraction and
+// injection of values, respectively.
+template <typename T>
+class B3Propagator : public B3PropagatorExtractor<T>
+{
+public:
+  // Rules that manages how context will be injected to carrier.
+  using Setter = void (*)(T &carrier,
+                          nostd::string_view trace_type,
+                          nostd::string_view trace_description);
+  // Sets the context for a HTTP header carrier with self defined rules.
+  void Inject(Setter setter, T &carrier, const context::Context &context) noexcept override
+  {
+    SpanContext span_context = B3PropagatorExtractor<T>::GetCurrentSpan(context);
+    if (!span_context.IsValid())
+    {
+      return;
+    }
+    char trace_id[32];
+    TraceId(span_context.trace_id()).ToLowerBase16(trace_id);
+    char span_id[16];
+    SpanId(span_context.span_id()).ToLowerBase16(span_id);
+    char trace_flags[2];
+    TraceFlags(span_context.trace_flags()).ToLowerBase16(trace_flags);
+    // Note: This is only temporary replacement for appendable string
+    std::string hex_string = "";
```

Comment:
This is on the hot path and will most likely allocate twice here when appending. Could use a stack buffer here with the appropriate size, since a string_view is passed to the setter anyway.